Translate faq-functions (#111)
reactjs/ja.reactjs.org: content/docs/faq-functions.md

}
```

> > **注意：**

Reviewer: ```suggestion
> **補足：**
```

method();
```

メソッドはバインドすることで、2 つ目のコードが 1 つ目と同様に動作するようになります。

> React では、一般的に他のコンポーネントに *渡す* メソッドしかバインドする必要はありません。たとえば、`<button onClick={this.handleClick}>` というコードは、`this.handleClick` を渡しているので、バインドする必要があります。しかし、レンダーやライフサイクルメソッドをバインドする必要はありません。それらは他のコンポーネントに渡すことがないからです。

Reviewer 2: ここはメソッド自体を指しているのでそのままの方がわかりやすいかなと感じました。 (imo)

```suggestion
React では、一般的に他のコンポーネントに *渡す* メソッドしかバインドする必要はありません。たとえば、`<button onClick={this.handleClick}>` というコードは、`this.handleClick` を渡しているので、バインドする必要があります。しかし、`render` メソッドやライフサイクルメソッドをバインドする必要はありません。それらは他のコンポーネントに渡すことがないからです。
```

> [Yefuda Kats の記事](http://yehudakatz.com/2011/08/11/understanding-javascript-function-invocation-and-this/) はバインドとは何か、また JavaScript における関数の動作について詳細に説明してくれています。

Reviewer 2: ```suggestion
[Yehuda Katz の記事](http://yehudakatz.com/2011/08/11/understanding-javascript-function-invocation-and-this/) はバインドとは何か、また JavaScript における関数の動作について詳細に説明してくれています。
```

> [Yefuda Kats によるこの記事](http://yehudakatz.com/2011/08/11/understanding-javascript-function-invocation-and-this/) で、バインドとは何か、また JavaScript における関数の動作について詳細に説明されています。

Reviewer 3: IMO: Thisは英語として成り立たせるための単語っぽいので訳さない方が自然な日本語になりそうです
1案ですが、以下を提案しておきます。
「Yefuda Kats の記事はバインドとは何か、また JavaScript における関数の動作について詳細に説明してくれています」

}
```

### 関数があまりに頻繁にあるいは連続して呼ばれる場合の対処法は？ {#how-can-i-prevent-a-function-from-being-called-too-quickly-or-too-many-times-in-a-row}

> `onClick` や `onScroll` のようなイベントハンドラがあって、それがあまりに高速に呼び出されるのを防ぎたい場合、コールバックが実行される頻度を制御するこができます。これは以下の方法で可能です：

Reviewer 3: IMO: 高速に呼び出されるだと、なぜ速度の話？ってなりそうなので、「連続して呼び出されるのを」ではいかがでしょう？

Reviewer: 「頻繁に」でもいいかも。

Author: @sasurau4 @smikitky ご意見ありがとうございます！確かに「速度」ではなく「頻度」が問題であり、さらにその程度が「過剰」であることが条件となっているため、より強調した「高頻度で」という表現にしたいと思います:pray:

- **スロットル**：時間ベースの頻度に基づくサンプルの制御（例：[`_.throttle`](https://lodash.com/docs#throttle)）
- **デバウンス**：一定時間アクティブでなかった場合の出力の制御（例：[`_.debounce`](https://lodash.com/docs#debounce))
- **`requestAnimationFrame` スロットル**：[`requestAnimationFrame`](https://developer.mozilla.org/en-US/docs/Web/API/window/requestAnimationFrame) に基づくサンプルの制御 (例：[`raf-schd`](https://github.com/alexreardon/raf-schd))

スロットルとデバウンスの比較は[この可視化](http://demo.nimius.net/debounce_throttle/) をご覧ください。

> > 注意：

Reviewer: ```suggestion
> 補足：
```

}
```

#### デバウンス {#debounce}

> デバウンスは、関数が最後に呼ばれてから一定時間経過するまで再実行されないようにします。これは、高速に発生するイベント（例：スクロールやキーボードイベントなど）に応じた高コストな計算処理が必要なときに役立ちます。下記の例は、250ミリ秒の遅延でテキスト入力をデバウンスします。

Reviewer: ```suggestion
デバウンスは、関数が最後に呼ばれてから一定時間経過するまで実行されないようにします。これは、高速に発生するイベント（例：スクロールやキーボードイベントなど）に応じた高コストな計算処理が必要なときに役立ちます。下記の例は、250ミリ秒の遅延でテキスト入力をデバウンスします。
```

「再」実行とは限らないです（ビジュアルデモ参照）

> デバウンスは、関数が最後に呼ばれてから一定時間経過するまで実行されないようにします。これは、高速に発生するイベント（例：スクロールやキーボードイベントなど）に応じた高コストな計算処理が必要なときに役立ちます。下記の例は、250 ミリ秒の遅延でテキスト入力をデバウンスします。

Reviewer 3: IMO: 案ですが「高速 」-> 「連続して」

Author: @sasurau4 先程の訳と足並みを揃えて「高頻度で」としてみます。

#### デバウンス {#debounce}

> デバウンスは、関数が最後に呼ばれてから一定時間経過するまで再実行されないようにします。これは、高頻度で発生するイベント（例：スクロールやキーボードイベントなど）に応じた高コストな計算処理が必要なときに役立ちます。下記の例は、250 ミリ秒の遅延でテキスト入力をデバウンスします。

Reviewer: 「再実行」の部分、「再」が復活してしまっています…

Author: @smikitky 失礼しました！修正します！

}
```

#### `requestAnimationFrame` スロットル {#requestanimationframe-throttling}

> [`requestAnimationFrame`](https://developer.mozilla.org/en-US/docs/Web/API/window/requestAnimationFrame) は描画パフォーマンスが最適化されるタイミングでブラウザで実行される関数をキューに入れる方法です。`requerstAnimationFrame` でキューに入れられている関数は、次のフレームで実行されます。ブラウザは、毎秒 60 フレーム (60fps) を保証するために懸命に働きます。しかし、それができなかった場合には、自然に 1 秒間のフレーム数が *制限* されます。たとえば、デバイスが 30 fps しか処理できなければ、1 秒間に 30 フレームしか取得できません。`requestAnimationFrame` スロットルを使うことで、1 秒間に 60 回以上になる更新を防ぐことができます。1 秒間に 100 回の更新をする場合、ユーザーには確認できない追加作業をブラウザが作成することになります。

Reviewer 3: IMO: ブラウザが一生懸命働きますだとちょっと微笑ましい感じがするので、「ブラウザは、毎秒** 60 フレーム (60fps) を確保するように努めます」はいかがでしょう？

> **注意：**
>

> > このテクニックにより、フレーム内の最後に公開された値のみがキャプチャされます。この最適化が動作する例は [`MDN`](https://developer.mozilla.org/en-US/docs/Web/Events/scroll) で確認できます。

Reviewer: ```suggestion
> このテクニックにより、フレーム内の最後に公開された値のみがキャプチャされます。この最適化がどのように動作するのかについての例は [`MDN`](https://developer.mozilla.org/en-US/docs/Web/Events/scroll) で確認できます。
```
